io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -266,3 +332,138 @@ ppEvents events =
 --
 runSimTrace :: forall a. (forall s. IOSim s a) -> SimTrace a
 runSimTrace mainAction = runST (runSimTraceST mainAction)
+
+controlSimTrace :: forall a. Maybe Int -> ScheduleControl -> (forall s. IOSim s a) -> SimTrace a
+controlSimTrace limit control mainAction = runST (controlSimTraceST limit control mainAction)
+
+exploreSimTrace ::
+  forall a test. (Testable test) =>
+    (ExplorationOptions->ExplorationOptions) ->
+    (forall s. IOSim s a) -> (Maybe (SimTrace a) -> SimTrace a -> test) -> Property
+exploreSimTrace optsf mainAction k =
+  case explorationReplay opts of
+    Nothing ->
+      explore (explorationScheduleBound opts) (explorationBranching opts) ControlDefault Nothing .&&.
+      let size = cacheSize() in size `seq`
+      tabulate "Modified schedules explored" [bucket size] True
+    Just control ->
+      replaySimTrace opts mainAction control k
+  where
+    opts = optsf stdExplorationOptions
+
+    explore n m control passingTrace =
+
+      -- ALERT!!! Impure code: readRaces must be called *after* we have
+      -- finished with trace.
+      let (readRaces,trace0) = detachTraceRaces $
+                                 controlSimTrace (explorationStepTimelimit opts) control mainAction
+          (sleeper,trace) = compareTraces passingTrace trace0
+      in (counterexample ("Schedule control: " ++ show control) $
+          counterexample (case sleeper of Nothing -> "No thread delayed"
+                                          Just ((t,tid,lab),racing) ->
+                                            showThread (tid,lab) ++
+                                            " delayed at time "++
+                                            show t ++
+                                            "\n  until after:\n" ++
+                                            unlines (map (("    "++).showThread) $ Set.toList racing)
+                                            ) $
+          k passingTrace trace) .&&|
+         let limit     = (n+m-1) `div` m
+             -- To ensure the set of schedules explored is deterministic, we filter out
+             -- cached ones *after* selecting the children of this node.
+             races     = filter (not . cached) . take limit $ readRaces()
+             branching = length races
+         in -- tabulate "Races explored" (map show races) $
+            tabulate "Branching factor" [bucket branching] $
+            tabulate "Race reversals per schedule" [bucket (raceReversals control)] $
+            conjoinPar
+              [ --Debug.trace "New schedule:" $
+                --Debug.trace ("  "++show r) $
+                --counterexample ("Schedule control: " ++ show r) $
+                explore n' ((m-1) `max` 1) r (Just trace0)
+              | (r,n') <- zip races (divide (n-branching) branching) ]
+
+    bucket n | n<10  = show n
+             | n>=10 = buck n 1
+             | otherwise = error "Ord Int is not a total order!"  -- GHC made me do it!
```

Comment:
Yes. I find the code clearer with the n>=10 guard in there too... I'd prefer to omit the otherwise clause. I'm with Dijkstra on this one (remember Dijkstra guarded commands?). Especially in the definition of buck, n>=10 is a necessary precondition for the following code to make sense. But maybe it's just me being a dinosaur!